pkg/test/framework/components/echo/kube/kube.go

```diff
@@ -0,0 +1,469 @@
+		"NoSidecar":            c.cfg.NoSidecar,
+		"IsHeadless":           c.cfg.IsHeadless,
+		"CreateServiceAccount": c.cfg.CreateServiceAccount,
+		"Ports":                c.cfg.Ports,
+	}
+
+	// Generate the YAML content.
+	var filled bytes.Buffer
+	w := bufio.NewWriter(&filled)
+	if err := deploymentTemplate.Execute(w, params); err != nil {
+		return nil, err
+	}
+	if err := w.Flush(); err != nil {
+		return nil, err
+	}
+
+	// Deploy the YAML.
+	d := deployment.NewYamlContentDeployment(c.cfg.Namespace.Name(), filled.String())
+	if err = d.Deploy(c.env.Accessor, false); err != nil {
+		return nil, err
+	}
+
+	c.deployment = d
+	return c, nil
+}
+
+func (c *instance) ID() resource.ID {
+	return c.id
+}
+
+func (c *instance) WaitUntilReady() error {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+
+	return c.init()
+}
+
+func (c *instance) init() error {
+	if c.client != nil {
+		// Already ready.
+		return nil
+	}
+
+	pod, err := c.waitUntilPodIsReady()
+	if err != nil {
+		return err
+	}
+
+	service, err := c.env.GetService(c.cfg.Namespace.Name(), c.cfg.Service)
+	if err != nil {
+		return err
+	}
+
+	// Extract the endpoints from the service definition.
+	c.endpoints = c.newEndpoints(*service, pod)
+
+	var grpcPort uint16
+	grpcPort, err = c.getGrpcPort()
+	if err != nil {
+		return err
+	}
+
+	// Create a forwarder to the command port of the app.
+	c.forwarder, err = c.env.NewPortForwarder(&testKube.PodSelectOptions{
+		PodNamespace: pod.Namespace,
+		PodName:      pod.Name,
+	}, 0, grpcPort)
+	if err != nil {
+		return err
+	}
+	if err = c.forwarder.Start(); err != nil {
+		return err
+	}
+
+	c.client, err = appEcho.NewClient(c.forwarder.Address())
+	if err != nil {
+		_ = c.forwarder.Close()
+		return err
+	}
+	return nil
+}
+
+func (c *instance) WaitUntilReadyOrFail(t testing.TB) {
+	if err := c.WaitUntilReady(); err != nil {
+		t.Fatal(err)
+	}
+}
+
+func (c *instance) Close() (err error) {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+
+	if c.client != nil {
+		err = c.client.Close()
+	}
+	if c.forwarder != nil {
+		err = multierror.Append(err, c.forwarder.Close()).ErrorOrNil()
+	}
+	if c.deployment != nil {
+		err = multierror.Append(err, c.deployment.Delete(c.env.Accessor, false))
+		err = multierror.Append(err, c.waitUntilPodIsDeleted()).ErrorOrNil()
+	}
+	return
+}
+
+func (c *instance) Config() echo.Config {
+	return c.cfg
+}
+
+func (c *instance) Endpoints() []echo.Endpoint {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+
+	// If we haven't already initialized the client, do so now.
+	if err := c.init(); err != nil {
+		log.Warnf("failed waiting for Endpoints: %v", err)
+		return make([]echo.Endpoint, 0)
+	}
+
+	out := make([]echo.Endpoint, len(c.endpoints))
+	for i, e := range c.endpoints {
+		out[i] = e
+	}
+	return out
+}
+
+func (c *instance) EndpointsForProtocol(protocol model.Protocol) []echo.Endpoint {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+
+	// If we haven't already initialized the client, do so now.
```

Comment:
Ah good question.  So we have a problem with k8s in that it takes time to start every pod ... could be 30 sec or more. The old apps component deployed all the apps first and then waited on them all in parallel.

Switching over to echo... If we do the wait in New, then the waits stack up and the test takes forever.  To avoid this, we have to build it into the interface. The test would create all of our Echo components, then wait until they're all ready.